File: Runtime/Native/Utils/Serializer/UserProfileSerializer.cs
using System.Collections.Generic;
using Io.AppMetrica.Internal;
using Io.AppMetrica.Internal.Profile;
using Io.AppMetrica.Profile;
using JetBrains.Annotations;
using System.Linq;

namespace Io.AppMetrica.Native.Utils.Serializer {
    internal static class UserProfileSerializer {
        [NotNull]
        public static string ToJsonString([NotNull] this UserProfile self) {
            return JSONEncoder.Encode(
                self.UserProfileUpdates
                    .Select(ConvertUpdate)
                    .Where(updateDict => updateDict != null)
                    .ToList()
            );
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertUpdate([NotNull] UserProfileUpdate value) {
            return ConvertBirthDate(value)
                   ?? ConvertBoolean(value)
                   ?? ConvertCounter(value)
                   ?? ConvertGender(value)
                   ?? ConvertName(value)
                   ?? ConvertNotificationsEnabled(value)
                   ?? ConvertNumber(value)
                   ?? ConvertString(value);
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertBirthDate(UserProfileUpdate value) {
            switch (value) {
                case BirthDateAgeUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BirthDateAgeUserProfileUpdate" },
                        { "Age", profileUpdate.Age },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case BirthDateYearUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BirthDateYearUserProfileUpdate" },
                        { "Year", profileUpdate.Year },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case BirthDateMonthUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BirthDateMonthUserProfileUpdate" },
                        { "Year", profileUpdate.Year },
                        { "Month", profileUpdate.Month },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case BirthDateDaysUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BirthDateDaysUserProfileUpdate" },
                        { "Year", profileUpdate.Year },
                        { "Month", profileUpdate.Month },
                        { "DayOfMonth", profileUpdate.DayOfMonth },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case BirthDateResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BirthDateResetUserProfileUpdate" },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertBoolean(UserProfileUpdate value) {
            switch (value) {
                case BooleanValueUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BooleanValueUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                        { "Value", profileUpdate.Value },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case BooleanResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "BooleanResetUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertCounter(UserProfileUpdate value) {
            switch (value) {
                case CounterDeltaUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "CounterDeltaUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                        { "Delta", profileUpdate.Delta },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertGender(UserProfileUpdate value) {
            switch (value) {
                case GenderValueUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "GenderValueUserProfileUpdate" },
                        { "Value", GetGenderString(profileUpdate.Value) },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case GenderResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "GenderResetUserProfileUpdate" },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static string GetGenderString(GenderAttribute.Gender? value) {
            switch (value) {
                case GenderAttribute.Gender.Female:
                    return "Female";
                case GenderAttribute.Gender.Male:
                    return "Male";
                case GenderAttribute.Gender.Other:
                    return "Other";
                default:
                    return null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertName(UserProfileUpdate value) {
            switch (value) {
                case NameValueUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "NameValueUserProfileUpdate" },
                        { "Value", profileUpdate.Value },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case NameResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "NameResetUserProfileUpdate" },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertNotificationsEnabled(UserProfileUpdate value) {
            switch (value) {
                case NotificationsEnabledValueUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "NotificationsEnabledValueUserProfileUpdate" },
                        { "Value", profileUpdate.Value },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case NotificationsEnabledResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "NotificationsEnabledResetUserProfileUpdate" },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertNumber(UserProfileUpdate value) {
            switch (value) {
                case NumberValueUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "NumberValueUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                        { "Value", profileUpdate.Value },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case NumberResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "NumberResetUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }

        [CanBeNull]
        private static IDictionary<string, object> ConvertString(UserProfileUpdate value) {
            switch (value) {
                case StringValueUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "StringValueUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                        { "Value", profileUpdate.Value },
                        { "IfUndefined", profileUpdate.IfUndefined },
                    };
                case StringResetUserProfileUpdate profileUpdate:
                    return new Dictionary<string, object> {
                        { "Type", "StringResetUserProfileUpdate" },
                        { "Key", profileUpdate.Key },
                    };
                default:
                    return (IDictionary<string, object>)null;
            }
        }
    }
}

File: Runtime/Plugins/Android/UserProfileSerializer.java
package io.appmetrica.analytics.plugin.unity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import io.appmetrica.analytics.profile.Attribute;
import io.appmetrica.analytics.profile.BirthDateAttribute;
import io.appmetrica.analytics.profile.BooleanAttribute;
import io.appmetrica.analytics.profile.CounterAttribute;
import io.appmetrica.analytics.profile.GenderAttribute;
import io.appmetrica.analytics.profile.NameAttribute;
import io.appmetrica.analytics.profile.NotificationsEnabledAttribute;
import io.appmetrica.analytics.profile.NumberAttribute;
import io.appmetrica.analytics.profile.StringAttribute;
import io.appmetrica.analytics.profile.UserProfile;
import io.appmetrica.analytics.profile.UserProfileUpdate;

final class UserProfileSerializer {
    private UserProfileSerializer() {}

    @NonNull
    public static UserProfile fromJsonString(@NonNull String profile) throws JSONException {
        JSONArray json = new JSONArray(profile);
        UserProfile.Builder builder = UserProfile.newBuilder();

        for (int idx = 0; idx < json.length(); idx++) {
            builder.apply(parseUserProfileUpdate(json.getJSONObject(idx)));
        }

        return builder.build();
    }

    @NonNull
    private static UserProfileUpdate<?> parseUserProfileUpdate(@NonNull JSONObject json) throws JSONException {
        String type = json.getString("Type");
        if (type.startsWith("BirthDate")) {
            return parseBirthDate(type, json);
        }
        if (type.startsWith("Boolean")) {
            return parseBoolean(type, json);
        }
        if (type.startsWith("Counter")) {
            return parseCounter(type, json);
        }
        if (type.startsWith("Gender")) {
            return parseGender(type, json);
        }
        if (type.startsWith("Name")) {
            return parseName(type, json);
        }
        if (type.startsWith("NotificationsEnabled")) {
            return parseNotificationsEnabled(type, json);
        }
        if (type.startsWith("Number")) {
            return parseNumber(type, json);
        }
        if (type.startsWith("String")) {
            return parseString(type, json);
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseBirthDate(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        BirthDateAttribute att = Attribute.birthDate();
        // BirthDate
        switch (type) {
            case "BirthDateAgeUserProfileUpdate": {
                int age = json.getInt("Age");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withAgeIfUndefined(age) : att.withAge(age);
            }
            case "BirthDateYearUserProfileUpdate": {
                int year = json.getInt("Year");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withBirthDateIfUndefined(year) : att.withBirthDate(year);
            }
            case "BirthDateMonthUserProfileUpdate": {
                int year = json.getInt("Year");
                int month = json.getInt("Month");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withBirthDateIfUndefined(year, month) : att.withBirthDate(year, month);
            }
            case "BirthDateDaysUserProfileUpdate": {
                int year = json.getInt("Year");
                int month = json.getInt("Month");
                int days = json.getInt("DayOfMonth");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withBirthDateIfUndefined(year, month, days) : att.withBirthDate(year, month, days);
            }
            case "BirthDateResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseBoolean(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        BooleanAttribute att = Attribute.customBoolean(json.getString("Key"));
        // BirthDate
        switch (type) {
            case "BooleanValueUserProfileUpdate": {
                boolean value = json.getBoolean("Value");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withValueIfUndefined(value) : att.withValue(value);
            }
            case "BooleanResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseCounter(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        CounterAttribute att = Attribute.customCounter(json.getString("Key"));
        // BirthDate
        if (type.equals("CounterDeltaUserProfileUpdate")) {
            double delta = json.getDouble("Delta");
            return att.withDelta(delta);
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseGender(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        GenderAttribute att = Attribute.gender();
        // BirthDate
        switch (type) {
            case "GenderValueUserProfileUpdate": {
                GenderAttribute.Gender value = getGender(json.getString("Value"));
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withValueIfUndefined(value) : att.withValue(value);
            }
            case "GenderResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static GenderAttribute.Gender getGender(@NonNull String value) {
        switch (value) {
            case "Female":
                return GenderAttribute.Gender.FEMALE;
            case "Male":
                return GenderAttribute.Gender.MALE;
            default:
                return GenderAttribute.Gender.OTHER;
        }
    }

    @NonNull
    private static UserProfileUpdate<?> parseName(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        NameAttribute att = Attribute.name();
        // BirthDate
        switch (type) {
            case "NameValueUserProfileUpdate": {
                String value = json.getString("Value");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withValueIfUndefined(value) : att.withValue(value);
            }
            case "NameResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseNotificationsEnabled(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        NotificationsEnabledAttribute att = Attribute.notificationsEnabled();
        // BirthDate
        switch (type) {
            case "NotificationsEnabledValueUserProfileUpdate": {
                boolean value = json.getBoolean("Value");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withValueIfUndefined(value) : att.withValue(value);
            }
            case "NotificationsEnabledResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseNumber(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        NumberAttribute att = Attribute.customNumber(json.getString("Key"));
        // BirthDate
        switch (type) {
            case "NumberValueUserProfileUpdate": {
                double value = json.getDouble("Value");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withValueIfUndefined(value) : att.withValue(value);
            }
            case "NumberResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }

    @NonNull
    private static UserProfileUpdate<?> parseString(@NonNull String type, @NonNull JSONObject json) throws JSONException {
        StringAttribute att = Attribute.customString(json.getString("Key"));
        // BirthDate
        switch (type) {
            case "StringValueUserProfileUpdate": {
                String value = json.getString("Value");
                boolean ifUndefined = json.getBoolean("IfUndefined");
                return ifUndefined ? att.withValueIfUndefined(value) : att.withValue(value);
            }
            case "StringResetUserProfileUpdate": {
                return att.withValueReset();
            }
        }
        throw new JSONException("Unknown UserProfile type " + type);
    }
}
